Add TPM 2.0 support for wolfSSH as a Client

tomoveu opened the pull request:
Intention of this PR is to protect the Private RSA Key of a wolfSSH Client

To achieve this I have gone throw several iterations and the best solution I found is to pass just the TPM key handle to SSH internals. Everything else, like wolfTPM init, TPM key loading and similar, should happen in the Client's code.

Note: TPM Handle is like a pointer to the TPM, which loaded key to use. Once we have this inside the SSH internals, it is clean to call a wolfTPM wrapper. It took me a while, to give up on deeper wolfTPM<>wolfSSH integration. Just the handle is simple and clean.

Project integration - I tried to use the power of automake through _AC_CONFIG_SUBDIRS_. Does not seem to yield complete success as header includes are not found.

Please bear with me @ejohnstown and @JacobBarthelmeh , as this is my first wolfSSH PR and this integration effort was not as straightforward as it may seem.

danielinux commented on wolfssh/ssh.h:
```diff
@@ -247,6 +247,7 @@ enum WS_FormatTypes {
     WOLFSSH_FORMAT_PEM,
     WOLFSSH_FORMAT_RAW,
     WOLFSSH_FORMAT_SSH,
+    WOLFSSH_FORMAT_TPM,
```
This WOLFSSH_FORMAT_TPM is not used in the client example. Did you plan to make an option to load the key from TPM via wolfSSH_ReadKey_* function ?

tomoveu replied:
This is a leftover from one of the previous tries (code iterations). Thanks for catching it.

Indeed, this was its use. I think this was variant 2 of the integration.

dgarske commented on examples/client/client.c:
```diff
@@ -792,6 +797,66 @@ static int wolfSSH_AGENT_IO_Cb(WS_AgentIoCbAction action,
+    }
+
+    /* TPM 2.0 keys live under a Primary Key, acquire such key */
+    if (getPrimaryStoragekey(dev, &storage, TPM_ALG_RSA)) {
```
We should only be using real WolfTPM public API's. Not our test key wrappers. Even if you need to move some of the code from examples/tpm_test_keys.c` into wolfSSH.

tomoveu replied:
I was wondering this. Will do.

tomoveu replied:
What to do with the IO Callbacks that are in examples/tpm_io.c(h) ?

I could just use NULL and have native Windows, Linux and TPM Simulator supports. And for baremetal to point to wolfTPM code.

dgarske replied:
I don't mind if you use the `example/tpm_io.c` for the other IO callbacks.

dgarske replied:
I would like to see all the code here calling wolfTPM native API's only. Bring in the `getPrimaryStoragekey` and `readKeyBlob` logic here. Eliminate the other files. I would like one copy/paste section of code as a template for this.

tomoveu commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "wolfTPM"]
+	path = wolfTPM
+	url = https://github.com/wolfSSL/wolfTPM.git
```
I think wolfBoot directly compiles from wolfTPM and adds the wolfTPM sources files as object files to its makefile using options.mk. I will move the submodule to the lib folder.

tomoveu commented on examples/client/client.c:
```diff
@@ -55,6 +55,11 @@
 #endif
 
+#ifdef WOLFSSH_TPM
+    #include "../../wolfTPM/wolftpm/tpm2_wrap.h"
```
OK. I will make the change.

tomoveu commented:
Changes:

* Fixed errors in code's logic
* wolfTPM is now used as a system library, same way wolfCrypt is used.
* removed wolfTPM as a git submodule (no longer needed)
* wolfTPM test helper functions are copied into examples/tpm and used in client's example, only if WOLFSSH_TPM is enabled
* Added set methods for tpm device and key

dgarske commented on configure.ac:
```diff
@@ -56,6 +56,7 @@ AC_CHECK_HEADERS([sys/select.h sys/time.h pty.h util.h termios.h])
 AC_CHECK_LIB([network],[socket])
 AC_CHECK_LIB([util],[forkpty])
 AC_CHECK_LIB([wolfssl],[wolfCrypt_Init],,[AC_MSG_ERROR([libwolfssl is required for ${PACKAGE}. It can be obtained from https://www.wolfssl.com/download.html/ .])])
+AC_CHECK_LIB([wolftpm],[wolfTPM2_Init],,[AC_MSG_ERROR([libwolftpm is required for ${PACKAGE}. It can be obtained from https://www.wolfssl.com/download.html/ .])])
```
This is being required all the time. It needs to be in the `AC_ARG_ENABLE([tpm],` section.

tomoveu replied:
This check is moved under the conditional check and now the build fails with ... when "--enable-tpm" is set.
```
Undefined symbols for architecture x86_64:
  "_wolfTPM2_SignHashScheme", referenced from:
      _SendKexDhReply in libwolfssh_la-internal.o
      _SendUserAuthRequest in libwolfssh_la-internal.o
```

Any idea why autoconf behaves like this?

dgarske commented on examples/tpm/tpm_key_helpers.c:
```diff
@@ -0,0 +1,361 @@
+/* helpers.c
```
Please correct comment name. Please add `#ifdef WOLFSSH_TPM` around all this. Please remove and cleanup for functions that are not used.

dgarske commented on examples/tpm/tpm_key_helpers.h:
```diff
@@ -0,0 +1,71 @@
+/* tpm_test_keys.h
```
Please correct comment name. Please add `#ifdef WOLFSSH_TPM` around all this. Please remove and cleanup for functions that are not used.

dgarske commented on examples/tpm/tpm_test.h:
```diff
@@ -0,0 +1,514 @@
+    extern "C" {
+#endif
+
+/* Test Configuration */
```
Please add `#ifdef WOLFSSH_TPM` around all this. Please cleanup and remove items not used for this example. A huge portion of this is not applicable.

dgarske commented on src/internal.c:
```diff
@@ -8147,9 +8156,16 @@ static int BuildUserAuthRequestRsa(WOLFSSH* ssh,
                 int sigSz;
                 WLOG(WS_LOG_INFO, "Signing hash with RSA.");
+#ifdef WOLFSSH_TPM
+                wolfTPM2_SignHashScheme(ssh->ctx->tpmDev, ssh->ctx->tpmKey,
```
I don't see any return code checking on this call? What if there is a TPM failure? What if `ssh->ctx->tpmDev` or `ssh->ctx->tpmKey` are not set?

dgarske commented on src/internal.c:
```diff
@@ -7109,8 +7112,14 @@ int SendKexDhReply(WOLFSSH* ssh)
                     WLOG(WS_LOG_INFO, "Signing hash with %s.",
                             IdToName(ssh->handshake->pubKeyId));
+#ifdef WOLFSSH_TPM
+                    wolfTPM2_SignHashScheme(ssh->ctx->tpmDev, ssh->ctx->tpmKey,
```
I don't see any return code checking on this call? What if there is a TPM failure? What if `ssh->ctx->tpmDev` or `ssh->ctx->tpmKey` are not set?

dgarske commented on src/ssh.c:
```diff
@@ -360,6 +360,26 @@ const char* wolfSSH_ErrorToName(int err)
+    WLOG(WS_LOG_DEBUG, "Entering wolfSSH_SetTpmKey()");
+
+    if (ssh)
+        ssh->ctx->tpmKey = key;
```
Might also check that `ssh->ctx` is not NULL.

dgarske commented on src/ssh.c:
```diff
@@ -360,6 +360,26 @@ const char* wolfSSH_ErrorToName(int err)
+    WLOG(WS_LOG_DEBUG, "Entering wolfSSH_SetTpmDev()");
+
+    if (ssh)
+        ssh->ctx->tpmDev = dev;
```
Might also check that `ssh->ctx` is not NULL.

dgarske commented on examples/tpm/tpm_key_helpers.h:
```diff
@@ -0,0 +1,71 @@
+#include <wolftpm/tpm2.h>
+#include <wolftpm/tpm2_wrap.h>
+
+WOLFTPM_LOCAL int readKeyBlob(const char* filename, WOLFTPM2_KEYBLOB* key);
```
The `WOLFTPM_LOCAL` is valid macro to use or have for the wolfSSH examples.

dgarske commented on examples/tpm/tpm_key_helpers.h:
```diff
@@ -0,0 +1,71 @@
+#ifndef _TPM_TEST_KEYS_H_
+#define _TPM_TEST_KEYS_H_
+
+#ifndef WOLFTPM2_NO_WRAPPER
```
All of the TPM support here requires `!WOLFTPM2_NO_WRAPPER`, so please make the build setting detection consistent for all TPM code. You can probably add a single well placed check:

```
#if defined(WOLFSSH_TPM) && defined(WOLFTPM2_NO_WRAPPER)
    /* wolfTPM support requires the wrapper to be enabled */
    #undef WOLFSSH_TPM
#endif
```

tomoveu replied:
Do you think this is required for the code in src/internal.c and ssh.c?

It is no longer required for examples/tpm/... of course, because the file is removed, and only the used code is in the client demo. I was wondering about src/internal.c and ssh.c

tomoveu commented:
@dgarske all feedback taken in, I have not finished testing. 

Also, I would like the ssh key generation in wolfTPM to be merged first, but there I have to resolve Key attributes issue first.

tomoveu commented:
I have two issues to resolve:

1. TPM issue -  turns out I need to use TPM2_RSA_Decrypt with manual padding to use TPM Key for SSH
2. Autoconf - moving the wolfTPM lib check under a conditional check creates one unresolved function

dgarske commented on examples/tpm/tpm_io.c:
```diff
@@ -0,0 +1,198 @@
+/* tpm_io.c
```
No, I do not want a copy of the example / reference tpm_io code in wolfSSH. That is duplicating a ton of code. This should reference the submodule example code...

tomoveu replied:
OK. Will fix.

dgarske commented on examples/client/client.c:
```diff
@@ -556,6 +560,133 @@ static inline void ato32(const byte* c, word32* u32)
+
+    XMEMSET(key, 0, sizeof(WOLFTPM2_KEYBLOB));
+
+    fp = XFOPEN(filename, "rb");
```
These should be using the wolfSSH port macros WFOPEN, WFSEEK, WFTELL, FREWIND, etc...

dgarske commented on examples/client/client.c:
```diff
@@ -55,6 +55,10 @@
 
+#ifdef WOLFSSH_TPM
+    #include <wolftpm/tpm2_wrap.h>
+    #include "../tpm/tpm_io.h"
```
Use the submodule examples/tpm_io.h

dgarske commented on examples/client/client.c:
```diff
@@ -556,6 +560,133 @@ static inline void ato32(const byte* c, word32* u32)
+    else {
+        /* specify auth password for storage key */
+        pStorageKey->handle.auth.size = sizeof(gStorageKeyAuth)-1;
+        XMEMCPY(pStorageKey->handle.auth.buffer, gStorageKeyAuth,
```
Use the wolfSSH macros WMEMCPY, etc...

dgarske commented on examples/client/client.c:
```diff
@@ -556,6 +560,133 @@ static inline void ato32(const byte* c, word32* u32)
+    byte pubAreaBuffer[sizeof(TPM2B_PUBLIC)];
+    int pubAreaSize;
+
+    XMEMSET(key, 0, sizeof(WOLFTPM2_KEYBLOB));
```
WMEMSET

dgarske commented on examples/client/client.c:
```diff
@@ -1017,6 +1228,15 @@ THREAD_RETURN WOLFSSH_THREAD client_test(void* args)
+     * This way, SSH client operations that can
+     * take advantage of the hardware security
+     * will have access to the TPM 2.0 device */
+    wolfSSH_SetTpmDev(ssh, &tpmDev);
```
Would a single API for both be easier?

tomoveu replied:
Initially, I thought like this also. Then, realized there would be an instance, when tpmDev is linked and only the tpmKey used is changed or has to be changed. 

Image the use case

Customer's wolfSSH app
- Init TPM
- Set tpmDEv
- Load Key
- Set tpmKey
- Use Key
- ......
- Unload handle, Load another Key
- Set **new** tpmKey
===== and we can not be sure tpmKey would be exposed in this scenario, so setting just the tpmKey makes sense.

IF, I add GetTpmDev, then Set could be a single call for both device and key

SetTpm(tpmDev, TpmKey)
... user operations ...
GetTpmDev(&tpmDev);
SetTpm(tpmDev, NewTpmKey);

dgarske reviewed (changes requested):
The .gitmodules is empty. Needs something like:

```
[submodule "lib/wolfTPM"]
	path = lib/wolfTPM
	url = https://github.com/wolfssl/wolfTPM
```

tomoveu commented:
@ejohnstown current status of this PR: I need the improvements/fixes to the Public Key authentication flow of wolfSSH to complete my work on TPM-backed Public Key authentication.

dgarske commented:
Closing to open new PR.
